### PR TITLE
feat: listStorageLocations optimization

### PR DESCRIPTION
**Pull request by bhudnell**
Description of changes:
Optimized `listStorageLocations` to use a `storageLocation` DDB entry rather than query and filter all `dataset` DDB entries.

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts [hunk 1]
```diff
@@ -122,7 +143,29 @@ export class DdbDataSetMetadataPlugin implements DataSetMetadataPlugin {
...
+      {
+        name: data.storageName as string,
+        type: data.storageType as string,
+        awsAccountId: data.awsAccountId as string,
+        region: data.region as string
+      },
+      'decrement'
+    );
+
+    if (remainingDatasets < 1) {
```
Comment (htztomic) [inline workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:161]: move this into _storeStorageLocationToDdb

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts [hunk 2]
```diff
@@ -122,7 +143,29 @@ export class DdbDataSetMetadataPlugin implements DataSetMetadataPlugin {
...
+      await this._aws.helpers.ddb
+        .delete({
+          pk: this._dataSetKeyType,
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:164]: shouldn't this be `buildDynamoDBPkSk(dataSetId, this._dataSetKeyType)`?

Comment (htztomic) [inline workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:164]: +1 ^ limit pk collision

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts [hunk 3]
```diff
@@ -240,6 +282,40 @@ export class DdbDataSetMetadataPlugin implements DataSetMetadataPlugin {
     });
   }
 
+  // Returns the current number of datasets for the storage location
+  private async _storeStorageLocationToDdb(
+    storageLocation: StorageLocation,
+    increment: 'increment' | 'decrement'
```
Comment (htztomic) [inline workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:288]:
```suggestion
    operator: 'increment' | 'decrement'
```

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts (hunk 2 above)
Comment (bhudnell) [reply workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:164]: storage location isnt unique to a specific dataset, so we cant use the dataset id in our `pk`/`sk`. I want 100% sure what to label as `pk` and `sk`. I ended up with the approach of `pk` being the dataset key type to signify that its part of the datasets entries, and `sk` to separate the different storage locations.

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts [hunk 4]
```diff
@@ -240,6 +282,40 @@ export class DdbDataSetMetadataPlugin implements DataSetMetadataPlugin {
...
+      sk: buildDynamoDbKey(storageLocation.name, this._storageLocationKeyType)
+    };
+
+    const { Attributes } = await this._aws.helpers.ddb.updateExecuteAndFormat({
+      key: storageLocationKey,
+      params: {
+        item: validatedStorageLocationMetadata.data,
+        set: '#datasetCount = if_not_exists(#datasetCount, :zero) + :counter',
+        names: { '#datasetCount': 'datasetCount' },
+        values: { ':counter': increment === 'increment' ? 1 : -1, ':zero': 0 }
```
Comment (htztomic) [inline workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:310]: in theory, you can end up with a dataset count of  < 0. Use a condition express to ensure no mitigate impact or have the condition commented above moved here to check for when it is < 1

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts (hunk 2 above)
Comment (htztomic) [reply workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:164]: understand, this should still be prefixed with something so you can limit potential collisions

Comment (bhudnell) [reply workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:164]: Would something like `pk: '${this._dataSetKeyType}#${this._storageLocationKeyType}'` work?

Comment (htztomic) [reply workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts:164]: if storageLocationKeyType is accessible during query time or doesn't impact your query patterns then yeah that is fine

File: workbench-core/example/infrastructure/integration-tests/support/complex/datasetHelper.ts [hunk 5]
```diff
@@ -50,14 +53,26 @@ export class DatasetHelper {
   }
 
   public async deleteDdbRecords(dataSetId: string): Promise<void> {
```
Comment (htztomic) [inline workbench-core/example/infrastructure/integration-tests/support/complex/datasetHelper.ts:55]: is this used for clean up?

Comment (bhudnell) [reply workbench-core/example/infrastructure/integration-tests/support/complex/datasetHelper.ts:55]: yes, in the `cleanup` task of the `Dataset` resource.

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts (hunk 1 above)
Comment (bhudnell) [reply workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts]: done

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts (hunk 3 above)
Comment (bhudnell) [reply workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts]: done

File: workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts (hunk 4 above)
Comment (bhudnell) [reply workbench-core/datasets/src/ddbDataSetMetadataPlugin.ts]: moved the "if count < 1, then delete" condition to this method.

**Review by DaveMadsen — APPROVED** [review]

**Review by aws-tyler — APPROVED** [review]

**Pull request merged**